### ISSUE TITLE
YUM repo

### ISSUE THREAD

**Issue by jtyr** [grafana/grafana#1668]
Please create a YUM repo for Grafana RPM packages. That will allow the users to upgrade to the latest version without the need to download the RPM package manually.

**Comment by torkelo** [comment]
Do you have any experience in doing this? Any pointers? Are there any public rpm hosting service that can be used?

**Comment by jtyr** [comment]
You can check this page:
https://www.digitalocean.com/community/tutorials/how-to-set-up-and-use-yum-repositories-on-a-centos-6-vps#custom-yum-repository

I don't know if there is any free hosting for YUM repos. People usually use AWS for that.

I have also inspected the RPM package and I do not see any dependencies on golang. Do we have to have `golang` installed to be able to run the Grafana backend? The way how the RPM package deals with the installation is not right. You should not symlink the init.d script or `/opt/grafana/versions/2.0.0-beta1` to `/opt/grafana/current`. RPM spec file should place the files where they belong. It should install everything into an unversioned directory (`/opt/grafana`) and let the RPM manage upgrades of this directory (remove files which are not in the newer version, override files which are coming from the newer version and differ from what's installed).

I can help you to create a proper RPM spec file which will allow to install Grafana in the right way on RedHat Enterprise Linux and on CentOS. Please just share with me the runtime requirements here.

**Comment by torkelo** [comment]
There are no dependencies on golang, I am using fpm to build deb & rpm packages. 

maybe the symlinking is not the usual way to do it. Is there a downside? I would be happy to receive help on this.

**Comment by jtyr** [comment]
I have created a SPEC file which allows to build RPM where files are distributed in standard directories provided by the OS on which the package is built:
https://github.com/picotrading/ansible-grafana/blob/master/contrib/grafana.spec

I have tested it only on CentOS 6 and 7 but it should work on RHEL and Fedora as well. We still need to create a systemd script and perhaps also a sysconfig file to allow to override variables specified in the init.d script. Please let me know if you are willing to use the SPEC file to build official Grafana RPMs and I will submit a PR for it.

**Comment by torkelo** [comment]
@jtyr thanks for taking the time to do this, it looks good. Will try it out today or tomorrow. I need to learn a little about the how spec files work and rpmbuild :) 

But looks better to have a real spec file than letting fpm generate one for me.

**Comment by torkelo** [comment]
@jtyr one question regarding package upgrades, is an ok practice to have a backup stage in the an upgrade script/step (would like to backup the grafana.db file for people who use the sqlite3 database with default location, or is this better handled some other way (update instructions, etc)). Grafana will automatically apply a SQL schema migration on startup  (there are currently no way to rollback the schema migrations), so a backup would be good.

**Comment by jtyr** [comment]
To test the SPEC file, you only need to do this:

1) install `rpm-build` package:
```
$ yum install rpm-build
```
2) On CentOS 6 you have to install newer version of Git:
curl -O ftp://ftp.pbone.net/mirror/ftp5.gwdg.de/pub/opensuse/repositories/home:/takehironet/CentOS_CentOS-6/x86_64/perl-Git-1.8.3.4-1.1.x86_64.rpm ftp://ftp.pbone.net/mirror/ftp5.gwdg.de/pub/opensuse/repositories/home:/takehironet/CentOS_CentOS-6/x86_64/git-1.8.3.4-1.1.x86_64.rpm

**Comment by jtyr** [comment]
Concerning the backup, I don't think so that it's a good idea to add any backup functionality into the RPM post installation script. Perhaps you could implement it into the Grafana backed to make backup automatically if the storage backend is SQLite. But I wouldn't bother even by this as it's user's responsibility to backup data before upgrading to newer version.

**Comment by torkelo** [comment]
@jtyr been working on new rpm and deb packages, see #1758  , no repo yet, but hope to do a repo soon

**Comment by rclsilver** [comment]
maybe you can create a repo in github?

**Comment by torkelo** [comment]
@rclsilver what do you mean?

**Comment by rclsilver** [comment]
I don't know that gitgub allow, but maybe there is a way to host some files (like an ftp server). In this case, you can use it as a yum repo (and aptitude repo too).

i'll make some research :)

**Comment by rclsilver** [comment]
In fact, download functionality is now deprecated as described here: https://github.com/blog/1302-goodbye-uploads

Sorry for the bad idea ^^

**Comment by torkelo** [comment]
@rclsilver hosting is not the problem. creating a rpm repo requires a lot more than just hosting, it is generating all the different files that define the repo and the signatures etc that will take time, and all those scripts needs to be automated into a CI process.

**Comment by rclsilver** [comment]
Yes, you're right, a yum repo is composed by a lot a files, generated by the createrepo command. 
But it's easy to "move" those files by FTP to publish them.
Regarding the process, i'm agree with you, the repo has to be built by your CI server.

**Comment by torkelo** [comment]
YUM repo, http://docs.grafana.org/installation/rpm/  (using Package Cloud  https://packagecloud.io/grafana)

**Comment by jtyr** [comment]
I'm really happy that Grafana is now available through YUM repo now. I only have couple of comments concerning the content of the RPM package:

* The service should not be enabled nor started from the post-installation script (that should be left for the user to do). Post-installation script should only add the `grafana` user.
* There is no `LICENSE` and `README` distributed with the package (they should be in `/usr/share/doc/grafana`).
* The `sample.ini` file should be in `/usr/share/doc/grafana` as it's part of the documentation (an example how to configure stuff).
* The main binary should be placed in `/usr/bin` and not in `/usr/sbin` (`sbin` should only be used for system management tools).
* There should not be any binary file in `/usr/share/grafana/bin` as it already exists in `/usr/bin`.
* You don't have to set `root:root` permissions on the `/etc/grafana/*` files in the post-installation script as that's already done by the RPM itself.
* You should not distribute `init.d` script with EL7 RPM, and you should not distribute `systemd` script with EL6 RPM.

**Comment by torkelo** [comment]
@jtyr thanks for your comments. 
1) Service should not be enabled nor started, this is already fixed in the new packages: https://github.com/grafana/grafana/blob/master/packaging/rpm/control/postinst
2) Will add the licence and readme, thanks for the pointer
3) I looked at a number of other server software rpms like RabbitMQ, Riak and Redis, all placed there binary in /usr/sbin

With seperating the EL7 & EL6 into seperate packages, I rather not do that as long as possible. Seems like pain to have to do separate packages for different dist versions. Elasticsearch rpm seems to follow that approach as well (distributes systemd and init.d scripts)

**Comment by jtyr** [comment]
Fair enough regarding the `sbin`. But please remove the redundant binary in `/usr/share/grafana/bin`. It only doubles the RPM size.

Regarding the same packages for different distro versions, that might become a bit tricky as you are building binary package, not `noarch`. Once the distro upgrades `glibc` then the RPM might become incompatible on the binary level and won't run nor install. It's strongly recommended to build binary RPMs on per-distro-version basis as only that guarantees the binary compatibility.

**Comment by torkelo** [comment]
@jtyr how do projects manage an automated build process that makes 50+ different packages for each combination of distro and version?

**Comment by jtyr** [comment]
They usually only create packages for the major distros - Red Hat and Debian. That brings a very good compatibility with their derivatives of the same version (RHEL 7 ~ CentOS 7 ~ SLC 7 ~ Fedora F19; Debian Sid_frozen_every_6_months ~ Ubuntu). Any other distro usually creates their own packages (Arch Linux, Gentoo Linux, ...).

If you have a spec file, you can spin up an instance in AWS with a specific OS and run the build on the VM. Once the build is done you can destroy the VM. Or you can have a set of VMs in Virtualbox which you use for the build. You don't even have to upgrade the VM because there are usually no major changes during the lifetime of the distro release.

**Comment by torkelo** [comment]
But if you need a separate build for each distro and version, 
so ubuntu/utopic, ubuntu/trusty, ubuntu/saucy, ubuntu/raring, ubuntu/quantal, ubuntu/precise
debian/lenny, debian/squeeze, debian/wheezy, debian/jessie, el/5, el/6, el/7, fedora/19, fedora/20, fedora/21, fedora/22 

That is a lot of packages, and VMs to spin up. Is there some service that can do this for you? Currently the packages are built in [this](https://github.com/grafana/grafana-packer) Centos 6.6 docker container, seems to create a binary that is compatible with most linux distros and versions I have tried.

**Comment by jtyr** [comment]
You can definitely use containers for that as well. Unfortunately I don't know about any such service which would do that for you. But you can be sure that the binary compatibility between distros will break after there will be major update in `glibc`. Packaging is not easy unless you have it automated.
